### DISCUSSION TITLE
Change SPI Instance

### DISCUSSION THREAD

**Discussion by geocheats2** [Networking-for-Arduino/EthernetENC#42] [General]
Do you have any idea how could i change the SPI instance from SPI to SPI_4 

I have tried pretty much everything and nothing seems to work

MCU STM32F446VE

Tried:

@ Enc28j60Network.cpp

`void Enc28J60Network::initSPI()
{
  if (spiInitialized)
    return;
  pinMode(csPin, OUTPUT);
  CSPASSIVE;
  SPI.setMISO(pinname)
  SPI.setMOSI(x)
  SPI.setSCLK(x)
  SPI.begin();
  spiInitialized = true;
}`

Tried redefining SPIClass SPI with on SPI.h with the pins of SPI_4

Tries changing the default pins of SPI @ variant.h

Ethernet Works Great @ SPI but not @ SPI_4

**Comment by JAndrassy** [comment]
I guess it is not OK to modify the default SPI instance.
but why not use the default SPI instance for Ethernet?

**Reply by geocheats2** [reply]
I got [this](https://github.com/FYSETC/FYSETC-SPIDER) board and the stepper motor driver header use SPI_4 while SPI is on SD Card

**Reply by JAndrassy** [reply]
but SPI is a bus. you can have SD card and Ethernet on the same SPI bus

**Reply by geocheats2** [reply]
you got me confused :P 
SD uses pins MISO PA6 / MOSI PA7/SCK PA5 while stepper motor driver headers use pins MISO PE13/MOSI PE14/SCK PE12
how can i use the same bus when i have different pins???

**Reply by JAndrassy** [reply]
put the enc28j60 on PA5/6/7 and use any other free pin as CS pin

**Reply by geocheats2** [reply]
i guess i could, but i have made a custom enc28j60 pcb board that uses the stepper driver headers and i would like to use one of those headers, i tried renaming all the instances of SPI.function in your library to a new SPIClass with the SPI_4 pins and it still didn't work i am starting to suspect that the documentation of the Spider Board i got is not correct. Thanks for your input :) 
Regarding your OTA library i did tried using it with platformio and of course the OTA.local it didn't work but the download from a server to flash worked perfectly, thanks again for your great work.

**Comment by geocheats2** [comment]
To anyone wondering how do to change the SPI instance of this library

!!! Warning tested only on STM32F4xxxxxxx !!!

Its not pretty but it works
1. Add the following after the #includes of Enc28J60Network.cpp
 SPIClass EthernetSPI(MOSI, MISO, SCLK);
 where MOSI , MISO, SCLK the pins you want to use
 !!!  Attention i had to use the actual pin numbers since the declaration wouldn't work with pin names !!! 
 eg SPIClass EthernetSPI(78, 77, 76);
2. Replace all instances of SPI.function with EthernetSPI.function
3. Compile & Run
4. ???
5. Profit

**Comment by JAndrassy** [comment]
why not `static SPIClass SPI(MOSI, MISO, SCLK);`?

or did you try `#define SPI SPI3`?

**Reply by geocheats2** [reply]
probably because i don't know how to use C++ properly :p

Thats why i said "Its not pretty but it works"

**Reply by geocheats2** [reply]
I did tried modifing Ethernet.init(CS,SPI); but i couldn't pass the spi instance to the library correctly
